runtime/src/mm/cpp/ObjectFactory.hpp

```diff
@@ -0,0 +1,317 @@
+                return;
+            }
+
+            std::lock_guard<SpinLock> guard(owner_.mutex_);
+
+            owner_.AssertCorrectUnsafe();
+
+            if (!owner_.root_) {
+                owner_.root_ = std::move(root_);
+            } else {
+                owner_.last_->next_ = std::move(root_);
+            }
+
+            owner_.last_ = last_;
+            last_ = nullptr;
+
+            RuntimeAssert(root_ == nullptr, "Must be empty");
+            AssertCorrect();
+            RuntimeAssert(owner_.root_ != nullptr, "Must not be empty");
+            owner_.AssertCorrectUnsafe();
+        }
+
+    private:
+        friend class ObjectFactoryStorage;
+
+        ALWAYS_INLINE void AssertCorrect() const noexcept {
+            if (root_ == nullptr) {
+                RuntimeAssert(last_ == nullptr, "last_ must be null");
+            } else {
+                RuntimeAssert(last_ != nullptr, "last_ must not be null");
+                RuntimeAssert(last_->next_ == nullptr, "last_ must not have next");
+            }
+        }
+
+        ObjectFactoryStorage& owner_; // weak
+        std::unique_ptr<Node> root_;
+        Node* last_ = nullptr;
+    };
+
+    class Iterator {
+    public:
+        Node& operator*() noexcept { return *node_; }
+        Node* operator->() noexcept { return node_; }
+
+        Iterator& operator++() noexcept {
+            previousNode_ = node_;
+            node_ = node_->next_.get();
+            return *this;
+        }
+
+        bool operator==(const Iterator& rhs) const noexcept { return node_ == rhs.node_; }
+
+        bool operator!=(const Iterator& rhs) const noexcept { return node_ != rhs.node_; }
+
+    private:
+        friend class ObjectFactoryStorage;
+
+        Iterator(Node* previousNode, Node* node) noexcept : previousNode_(previousNode), node_(node) {}
+
+        Node* previousNode_; // Kept for `Iterable::EraseAndAdvance`.
+        Node* node_;
+    };
+
+    class Iterable : private MoveOnly {
+    public:
+        explicit Iterable(ObjectFactoryStorage& owner) noexcept : owner_(owner), guard_(owner_.mutex_) {}
+
+        Iterator begin() noexcept { return Iterator(nullptr, owner_.root_.get()); }
+        Iterator end() noexcept { return Iterator(owner_.last_, nullptr); }
+
+        void EraseAndAdvance(Iterator& iterator) noexcept { iterator.node_ = owner_.EraseUnsafe(iterator.previousNode_); }
+
+    private:
+        ObjectFactoryStorage& owner_; // weak
+        std::unique_lock<SpinLock> guard_;
+    };
+
+    // Lock `ObjectFactoryStorage` for safe iteration.
+    Iterable Iter() noexcept { return Iterable(*this); }
+
+private:
+    // Expects `mutex_` to be held by the current thread.
+    Node* EraseUnsafe(Node* previousNode) noexcept {
+        RuntimeAssert(root_ != nullptr, "Must not be empty");
+        AssertCorrectUnsafe();
+
+        if (previousNode == nullptr) {
+            // Deleting the root.
+            auto newRoot = std::move(root_->next_);
```

Comment:
Why is it impossible to move it directy: `root_ = std::move(root_->next_)`?

Reply:
Fixed, thanks!